File: Python/Sorting/HeapSort.py
"""
Heap Sort Implementation

Time Complexity:
- Best Case: O(n log n)
- Worst Case: O(n log n)
- Average Case: O(n log n)

Space Complexity: O(1) - in-place sorting.
"""

def heap_sort(arr):
    n = len(arr)
    res = list(arr)

    # Build a maxheap.
    for i in range(n // 2 - 1, -1, -1):
        heapify(res, n, i)

    # One by one extract elements
    for i in range(n - 1, 0, -1):
        res[i], res[0] = res[0], res[i]  # swap
        heapify(res, i, 0)
    
    return res

def heapify(arr, n, i):
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2

    if l < n and arr[i] < arr[l]:
        largest = l

    if r < n and arr[largest] < arr[r]:
        largest = r

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

# Example usage:
# if __name__ == "__main__":
#     data = [5, 3, 8, 4, 2]
#     print(heap_sort(data))

File: CPP/Sorting/HeapSort.cpp
/**
 * Heap Sort Implementation
 * 
 * Time Complexity:
 * - Best Case: O(n log n)
 * - Worst Case: O(n log n)
 * - Average Case: O(n log n)
 * 
 * Space Complexity: O(1) - in-place sorting.
 */

#include <iostream>
#include <vector>
#include <algorithm>

template <typename T>
void heapify(std::vector<T>& arr, int n, int i) {
    int largest = i;
    int l = 2 * i + 1;
    int r = 2 * i + 2;

    if (l < n && arr[l] > arr[largest])
        largest = l;

    if (r < n && arr[r] > arr[largest])
        largest = r;

    if (largest != i) {
        std::swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

template <typename T>
void heapSort(std::vector<T>& arr) {
    int n = arr.size();

    // Build heap
    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(arr, n, i);

    // Extract elements from heap
    for (int i = n - 1; i > 0; i--) {
        std::swap(arr[0], arr[i]);
        heapify(arr, i, 0);
    }
}

/*
int main() {
    std::vector<int> data = {5, 3, 8, 4, 2};
    heapSort(data);
    for (int x : data) std::cout << x << " ";
    return 0;
}
*/
